ts2mls/js/src/main/scala/ts2mls/TSCompilerInfo.scala

```diff
@@ -0,0 +1,170 @@
+package ts2mls
+
+import scala.scalajs.js
+import js.Dynamic.{global => g}
+import js.DynamicImplicits._
+import js.JSConverters._
+import ts2mls.types._
+
+object TypeScript {
+  private val ts = g.require("typescript")
+
+  val typeFlagsEnumLike = ts.TypeFlags.EnumLike
+  val typeFlagsObject = ts.TypeFlags.Object
+  val typeFlagsTypeParameter = ts.TypeFlags.TypeParameter
+  val syntaxKindPrivate = ts.SyntaxKind.PrivateKeyword
+  val syntaxKindProtected = ts.SyntaxKind.ProtectedKeyword
+  val syntaxKindStatic = ts.SyntaxKind.StaticKeyword
+  val objectFlagsAnonymous = ts.ObjectFlags.Anonymous
+
+  def isToken(node: js.Dynamic) = ts.isToken(node)
+  def isClassDeclaration(node: js.Dynamic) = ts.isClassDeclaration(node)
+  def isInterfaceDeclaration(node: js.Dynamic) = ts.isInterfaceDeclaration(node)
+  def isFunctionLike(node: js.Dynamic) = ts.isFunctionLike(node)
+  def isModuleDeclaration(node: js.Dynamic) = ts.isModuleDeclaration(node)
+  def isArrayTypeNode(node: js.Dynamic) = ts.isArrayTypeNode(node)
+  def isTupleTypeNode(node: js.Dynamic) = ts.isTupleTypeNode(node)
+
+  def forEachChild(root: js.Dynamic, func: js.Dynamic => Unit) = ts.forEachChild(root, func)
+  def createProgram(filenames: Seq[String]) =
+    ts.createProgram(filenames.toJSArray, js.Dictionary("maxNodeModuleJsDepth" -> 0, "target" -> ts.ScriptTarget.ES5, "module" -> ts.ModuleKind.CommonJS))
+}
+
+class TSTypeChecker(checker: js.Dynamic) {
+  def getTypeStringOfSymbol(sym: js.Dynamic): String =
+    checker.typeToString(getTypeOfSymbolAtLocation(sym, sym.valueDeclaration)).toString
+
+  def getReturnTypeOfSignature(node: js.Dynamic) = checker.getReturnTypeOfSignature(checker.getSignatureFromDeclaration(node))
+  def getTypeFromTypeNode(node: js.Dynamic) = checker.getTypeFromTypeNode(node)
+  def getTypeOfSymbolAtLocation(sym: js.Dynamic, node: js.Dynamic) = checker.getTypeOfSymbolAtLocation(sym, node)
+  def getPropertiesOfType(tp: js.Dynamic) = checker.getPropertiesOfType(tp)
+
+  def isImplementationOfOverload(node: js.Dynamic) = checker.isImplementationOfOverload(node)
+  def typeToTypeNode(tp: js.Dynamic) = checker.typeToTypeNode(tp)
+  def getTypeArguments(tp: js.Dynamic) = checker.getTypeArguments(tp)
+  def getElementTypeOfArrayType(tp: js.Dynamic) = checker.getElementTypeOfArrayType(tp)
+}
+
+object TSTypeChecker {
+  def apply(checker: js.Dynamic) = new TSTypeChecker(checker)
+}
+
+class TSSymbolObject(sym: js.Dynamic)(implicit checker: TSTypeChecker) extends TSAny(sym) {
+  private lazy val parent = TSSymbolObject(sym.parent)
+
+  // the first declaration of this symbol
+  // if there is no overloading, there is only one declaration
+  lazy val declaration =
+    if (declarations.isUndefined) TSNodeObject(g.undefined)
+    else declarations.get(0)
+  lazy val escapedName: String = sym.escapedName.toString
+  lazy val declarations = TSNodeArray(sym.declarations)
+
+  lazy val symbolType: String = checker.getTypeStringOfSymbol(sym)
```

Comment:
Can you add some documentation? Why do we need the string of a type here?

Reply:
This name seems ambiguous. I've changed it and added some explains.

Reply:
I'm still quite confused by both the new name and the comment. What is the value of this thing supposed to be when the symbol does not correspond to a builtin type?!

Reply:
That would be a string of type in TypeScript style.
For example:
```ts
function add(x: number, y: number): number {
  return x + y
}
```
If we invoke this API, it would return `(x: number, y: number) => number`.

Reply:
Then the name is completely inappropriate and doesn't describe what it is at all ;)

I'd prefer if you used a specialized method in the specialized cases where this makes sense (such as directly accessing the builtin's name from the appropriate field), instead of polluting _all_ symbols with a method which doesn't make much sense for most of them.

Stringly-typed approaches, based on converting things to strings, are a recipe for unmaintainable code and disastrous maintenance.

Reply:
I used another approach to get the `TypeObject` of objects, and removed this ambiguous member. 😄